agent/api/task_unix.go

```diff
@@ -15,16 +15,119 @@
 
 package api
 
+import (
+	"path/filepath"
+
+	"github.com/aws/amazon-ecs-agent/agent/config"
+	"github.com/aws/amazon-ecs-agent/agent/utils"
+	"github.com/cihub/seelog"
+	docker "github.com/fsouza/go-dockerclient"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/pkg/errors"
+)
+
 const (
 	portBindingHostIP = "0.0.0.0"
+	maxTaskVCPULimit  = 10     // 10 VCPUs limit
+	defaultCPUPeriod  = 100000 // 100ms
+	// Reference: http://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDefinition.html
+	defaultCPUShare = 2
 )
 
-func (task *Task) adjustForPlatform() {}
+func (task *Task) adjustForPlatform(cfg *config.Config) {
+	task.memoryCPULimitsLock.Lock()
+	defer task.memoryCPULimitsLock.Unlock()
+	task.memoryCPULimits = cfg.TaskCPUMemLimit
+}
 
 func getCanonicalPath(path string) string { return path }
 
-// CgroupEnabled checks whether cgroups need be enabled at the task level
-func (task *Task) CgroupEnabled() bool {
-	//stub for now
-	return false
+// BuildCgroupRoot helps build the task cgroup prefix
+// Example: /ecs/task-id
+func (task *Task) BuildCgroupRoot() (string, error) {
+	taskID, err := task.GetID()
+	if err != nil {
+		return "", errors.Wrapf(err, "task build cgroup root: unable to get task-id from taskARN: %s", task.Arn)
+	}
+
+	cgroupRoot := filepath.Join(config.DefaultTaskCgroupPrefix, taskID)
+	return cgroupRoot, nil
+}
+
+// BuildLinuxResourceSpec returns a linuxResources object for the task cgroup
+func (task *Task) BuildLinuxResourceSpec() (specs.LinuxResources, error) {
+	linuxResourceSpec := specs.LinuxResources{}
+
+	if !utils.ZeroOrNil(task.VCPULimit) {
+		if task.VCPULimit > maxTaskVCPULimit || task.VCPULimit < 0 {
+			return specs.LinuxResources{},
+				errors.Errorf("task resource spec builder: unsupported CPU limits, requested=%f, max-supported=%d",
+					task.VCPULimit, maxTaskVCPULimit)
+		}
+		taskCPUPeriod := uint64(defaultCPUPeriod)
+		taskCPUQuota := int64(task.VCPULimit * defaultCPUPeriod)
+
+		// TODO: DefaultCPUPeriod only permits 10VCPUs.
+		// Adaptive calculation of CPUPeriod required for further support
+		linuxResourceSpec.CPU = &specs.LinuxCPU{
+			Quota:  &taskCPUQuota,
+			Period: &taskCPUPeriod,
+		}
+	} else {
+		var taskCPUShares uint64
+		for _, container := range task.Containers {
+			if !utils.ZeroOrNil(container.CPU) {
+				taskCPUShares += uint64(container.CPU)
```

Comment:
A comment in the code here would improve understandability

Reply:
Sure